omniledger/service/transaction.go

```diff
@@ -110,19 +110,48 @@ func (instr Instruction) Hash() []byte {
 // and the given string.
 //
 // DeriveID is used inside of contracts that need to create additional keys in
-// the collection. Newly spawned instances should, by convention, always use the
-// Hash() of the instruction as the new InstanceID.
+// the collection. By convention newly spawned instances should have their
+// InstanceID derived via inst.DeriveID("").
 func (instr Instruction) DeriveID(what string) InstanceID {
+	var b [4]byte
+
+	// Une petit primer on domain separation in hashing:
+	// Domain separation is required when the input has variable lengths,
+	// because an attacker could try to construct two messages resulting in the
+	// same hash by moving bytes from one neighboring input to the
+	// other. With fixed-length inputs, moving bytes is not possible, so
+	// no domain separation is needed.
+
 	h := sha256.New()
 	h.Write(instr.Hash())
-	h.Write([]byte{0})
+
+	// Putting the length of instr.Signatures into the hash does not seem to
```

Comment:
The reason we do this (according to Jeff ;) is because we might have OpenIDConnect signatures that have variable-length signatures and that might be prone to padding-attacks.